riscv/src/register/mcause.rs

```diff
@@ -30,10 +13,32 @@ pub enum Interrupt {
     MachineTimer = 7,
     SupervisorExternal = 9,
     MachineExternal = 11,
-    Unknown,
 }
 
-/// Exception
+/// SAFETY: `Interrupt` represents the standard RISC-V interrupts
+unsafe impl InterruptNumber for Interrupt {
+    const MAX_INTERRUPT_NUMBER: usize = Self::MachineExternal as usize;
+
+    #[inline]
+    fn number(self) -> usize {
+        self as usize
+    }
+
+    #[inline]
+    fn from_number(value: usize) -> Result<Self, usize> {
+        if value > 11 || value % 2 == 0 {
+            Err(value)
+        } else {
+            // SAFETY: valid interrupt number
+            unsafe { Ok(core::mem::transmute::<usize, Self>(value)) }
+        }
```

Comment:
Could we change this to a:

```rust
match value {
    1 => Ok(Self::SupervisorSoft),
    // ...
    _ => Err(value),
}
```

Mainly to avoid the call to `transmute`, since we really don't need it.

Reply:
The good thing about my `transmute` implementation is that on average it is faster (no need to have N cases, with just two comparisons you are ready to go)

Reply:
The bad thing about `transmute` is that it is inherently unsafe (read "can introduce UB"), and should generally be avoided whenever possible.

Also, `transmute` is not necessarily faster:

```
from_match              time:   [147.28 ps 147.82 ps 148.47 ps]
                        change: [-2.0146% -0.1924% +1.7735%] (p = 0.86 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) high mild
  9 (9.00%) high severe

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [150.79 ps 151.19 ps 151.64 ps]
                        change: [-0.1834% +2.5244% +4.7970%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe
```

Running multiple iterations of `cargo bench` yields some cases where `transmute` is faster, but only slightly. The two implementations swap which is faster, sometimes `match` sometimes `transmute`. Always +/- ~4-5ps.

`match` has the advantage of always being safe.

```rust
#[repr(usize)]
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum F {
    A = 1,
    B = 2,
    C = 3,
    D = 4,
}

impl F {
    pub fn from_num_transmute(val: usize) -> Option<Self> {
        if val == 0 || val > 4 {
            None
        } else {
            Some(unsafe { std::mem::transmute::<usize, Self>(val) })
        }
    }

    pub fn from_num_match(val: usize) -> Option<Self> {
        match val {
            1 => Some(Self::A),
            2 => Some(Self::B),
            3 => Some(Self::C),
            4 => Some(Self::D),
            _ => None,
        }
    }
}
```

The benchmark code:

```rust
// Using `match`
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ff::F;

fn bench_match(n: usize) {
    F::from_num_match(n);
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("from_match", |b| b.iter(|| bench_match(black_box(20))));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
```

```rust
// Using `transmute
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ff::F;

fn bench_transmute(n: usize) {
    F::from_num_transmute(n);
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("from_transmute", |b| b.iter(|| bench_transmute(black_box(20))));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
```
```rust

Reply:
This got me thinking about why the two implementations are so close in performance, and how the compiler translates to machine code.

Here is an example from [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATAHZSjgBk8BkwAOXcAI0xiCQAOUgAHVAVCOwYXNw9vPySU2wEgkPCWKJiueKtMGzShAiZiAgz3T19LTGt8hlr6gkKwyOi4yzqGpqzW8x6%2B4tK4gEpLVBNiZHYOAFIvAGYwMHWAVgAhBlQAfXN0A4ARdY0AQVuH7YPD4kwE4ggzPAAvTDnro8EiYIgBqRjuUHYUHrHyHR6gxGgu4wrZXUFcUgIpGHVHorZY%2B5I0FOPGg/aElFI9HrNGgvzYxHQ2nogCclOJADEyVxMY9YTd7vznkcTqcWIZgPRAfdgWCqAxQVRiKgWKcCMRDAoWCYCJgIAx3CBQd8/nNQQBaWnQgDyCU6tKc2GtMLhjNBeCooMNLFR0N5rqcsNJPphXn2oK8eJZoI0rvhROJSNCAkw7oF4NoSnj7uJQlV%2BpMDAUTComHjoLQbxAIDYLBrGq1Or1NcdpswpCh1oN7nNAoBicRAv5PkFDyFIuOZwlRml%2BzHcqViuVqvFTAIyAQPfrJpSZstLrtDq2TpdsITVMREo3CG9kPPuaRXGjW2h%2BbYEGwNbuC0fiK2L5vgWn41ocv6DkmEYsi6776l%2BIBOOBl5Jj4gGgrBIEgFcSFJkirJoRh8HYDhuEYs%2B0GvuhwHwZyJG4acaEpiEHJIsOQqjvy9wAPRcaCACSXoAJ5LLu5brAAbBoErBBAAKSZ2WDIAYbwegQoJMAoMKSXKElxkQoJKOWhAeoqBAIOWSwEMCBAgI8PGgouCqgtJDCyRWAB0nmumOHALLQnD7LwnjcLwqCcAASmYakKEsKxidsPCkDZHBaHMCwANYgFsWzudleX5QV4n6JwkhBZooWcLwCggBoSXlQscCwEgaAsAkdDROQlAtW19AxMQXA%2BLEtU0LQerENVEAROVpARME9SCZwiWzcwxCCTaETaFUyWJS1bCCDaDC0AtKW8FgEQmMAThiFmi2nZgM5SmsWikPgbzVAAbpg1Unf4qhVLqayJcEer%2BT9tB4BEmqrS4WDTRqeD1iFpCfcQETJJgVz3ZK4NGPVfAGMACgAGp4JgADudqMLdMiCCIYjsFINPyEoajTbomIGLjpjmPoEPVZACyoPaaTfRapKkRLQsEBa9CfbQLIARaFojMAmBqcQeAKMg73iZIwDIBaRbpScZMMBaJwhBa7RUKFKMa1g/OyW0HRpA4DDOK4zR6IEwT9CUgyYrkqQCGMniB8kwcMNMAxlM7W01CMjSe1kmKVNUAjdA00f%2B7HkyjMnYfDFMvszAHCwxcsqwSH5AVlT9YUcKC3Mbhi7mDe5cafvgRDEGGWxcHMvDJalGVZTlBUT3lRWg6VpDBc9DdVTVdUnQ1MCICAlnWZ1EDde1xChKwazN8grft89mDdyQeDoHo/C06I4iM/fzMqOoP3s6QZOagkt01xwgU57TQbjaXU1lQSoC9CfM%2BsQO6gggC4Vq%2B8%2B4DyHvVUe2VcqTwnsVDgs954VQ4EvWqw9fL/y8HXBelUV4j2RtEFI9hJBAA%3D%3D):

```rust
#![no_std]

#[repr(usize)]
pub enum E {
    A = 1,
    B = 3,
    C = 5,
    D = 7,
    E = 9,
    F = 11,
}

#[no_mangle]
pub fn from_transmute(num: usize) -> Option<E> {
    if num > 11 || num % 2 == 0 {
        None
    } else {
        Some(unsafe { core::mem::transmute::<usize, E>(num) })
    }
}

#[no_mangle]
pub fn from_match(num: usize) -> Option<E> {
    match num {
        1 => Some(E::A),
        3 => Some(E::B),
        5 => Some(E::C),
        7 => Some(E::D),
        9 => Some(E::E),
        11 => Some(E::F),
        _ => None,
    }
}
```

---

```asm
; with rustc -C opt-level=3 --target riscv64gc-unknown-none-elf
from_transmute:
        sltiu   a1, a0, 12
        xori    a1, a1, 1
        andi    a2, a0, 1
        seqz    a2, a2
        or      a1, a1, a2
        addi    a1, a1, -1
        and     a0, a0, a1
        ret

from_match:
        addi    a0, a0, -1
        li      a1, 10
        bltu    a1, a0, .LBB1_2
        slli    a0, a0, 3
.Lpcrel_hi0:
        auipc   a1, %pcrel_hi(.Lswitch.table.from_match)
        addi    a1, a1, %pcrel_lo(.Lpcrel_hi0)
        add     a0, a0, a1
        ld      a0, 0(a0)
        ret
.LBB1_2:
        li      a0, 0
        ret

.Lswitch.table.from_match:
        .quad   1
        .quad   0
        .quad   3
        .quad   0
        .quad   5
        .quad   0
        .quad   7
        .quad   0
        .quad   9
        .quad   0
        .quad   11
```